### PR TITLE
Feat/2fa email otp support

### LINKED ISSUE #1279: Allow 2FA activation/deactivation for social login users without password verification
### Is this suited for github?

- [x] Yes, this is suited for github

## Description
Currently, users who sign in through social providers (OAuth) cannot enable or disable Two-Factor Authentication (2FA) because the system requires password verification, which these users don't have.

## Current Behavior
- Users who sign in via social providers (Google, GitHub, etc.) can't activate 2FA
- The current implementation requires password verification for:
  - Enabling 2FA (`/two-factor/enable`)
  - Disabling 2FA (`/two-factor/disable`)
  - Generating new backup codes (`/two-factor/generate-backup-codes`)

## Expected Behavior
Users should be able to manage their 2FA settings regardless of their authentication method while maintaining security.

## Proposed Solution
Implement an alternative verification method for social login users:

1. **Email OTP Verification:**
   - When a social login user attempts to enable/disable 2FA:
     - Generate a one-time code
     - Send it to their verified email
     - Require this code for verification instead of password

```typescript
// Example implementation
interface TwoFactorVerification {
  type: 'password' | 'email_otp';
  value: string;
}



### Describe the solution you'd like


1. When social login users try to enable/disable 2FA:
   - Generate a one-time email verification code
   - Send it to their verified email
   - Use this code instead of password verification

2. Flow would be:
User -> Request 2FA Change -> System Detects Social Login or passwordless login ->
Send Email OTP -> User Verifies OTP -> Enable/Disable 2FA

### Describe alternatives you've considered

I dont see any safe other alternatives

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by helloworld9912**
# Add Email OTP Verification Support for 2FA Operations

## Overview
This PR adds support for email OTP verification across all Two-Factor Authentication operations, making 2FA more accessible for users who authenticate through social login providers or prefer email verification.

## Key Changes
- Added unified verification schema supporting both password and email OTP methods
- Implemented email OTP verification for:
  - Enabling/disabling 2FA
  - Generating backup codes
  - Retrieving TOTP URI
- Added comprehensive error handling for OTP verification scenarios
- Enhanced TypeScript types for better type safety and developer experience
- Added extensive test coverage for email OTP verification flows

## Implementation Details
- Created `twoFactorVerificationSchema` for consistent verification across endpoints
- Added `sendVerificationOTP` handler in plugin configuration
- Implemented proper OTP validation with expiration checks
- Added logging for verification attempts to aid debugging
- Updated API documentation to reflect new verification methods

## Testing
- Added test cases for all email OTP verification scenarios
- Verified compatibility with existing password-based verification
- Tested error handling and edge cases
- Verified social login user flows

## Related
- Plugin configuration documentation will be updated in a separate PR
- Fixes #1279 

## Checklist
- [x] Added TypeScript types
- [x] Added test coverage
- [x] Updated error messages
- [x] Added logging
- [x] Tested with social login flows
- [x] Maintained backward compatibility
- [x] Updated docs

### CODE REVIEW

**Comment by GregorWedlich** [comment]
Oh yes, that looks good, we need this function as soon as possible :)

File: docs/content/docs/plugins/2fa.mdx [hunk 1]
```diff
@@ -80,11 +80,28 @@ This plugin offers two main methods to do a second factor verification:
...
 
-To enable two-factor authentication, call `twoFactor.enable` with the user's password:
+
+To enable two-factor authentication, you can use either password verification or email OTP verification (useful for social login users):
 
 ```ts title="two-factor.ts"
+// Using password verification
 const { data } = await authClient.twoFactor.enable({
-    password: "password" // user password required
+    verification: {
```
Comment (Bekacru) [inline docs/content/docs/plugins/2fa.mdx]: this would be a breaking change. So instead of `verification` let's just make the body type a union of `password` or `otp` instead.

File: packages/better-auth/src/plugins/two-factor/types.ts [hunk 2]
```diff
@@ -33,6 +33,18 @@ export interface TwoFactorOptions {
...
 	schema?: InferOptionSchema<typeof schema>;
+	/**
+	 * Function to send email verification for 2FA changes
+	 * If not provided, will fallback to password verification
+	 */
+	sendVerificationOTP?: (
+		data: {
+			email: string;
+			otp: string;
+			type: "2fa-enable" | "2fa-disable";
```
Comment (Bekacru) [inline packages/better-auth/src/plugins/two-factor/types.ts]: let's rename this to `status` and the values should be `enable` or `disable`

File: packages/better-auth/src/plugins/two-factor/index.ts [hunk 3]
```diff
@@ -18,6 +18,44 @@ import { BASE_ERROR_CODES } from "../../error/codes";
 import { createOTP } from "@better-auth/utils/otp";
 import { createHMAC } from "@better-auth/utils/hmac";
 
+export const twoFactorVerificationSchema = z.object({
```
Comment (Bekacru) [inline packages/better-auth/src/plugins/two-factor/index.ts]: let's remove `verification` and they should only pass between `password` and `otp`. No `type` parameter as well.

File: packages/better-auth/src/plugins/two-factor/index.ts [hunk 4]
```diff
@@ -36,19 +74,13 @@ export const twoFactor = (options?: TwoFactorOptions) => {
...
 				{
 					method: "POST",
-					body: z.object({
-						password: z
-							.string({
-								description: "User password",
-							})
-							.min(8),
-					}),
+					body: twoFactorVerificationSchema,
```
Comment (Bekacru) [inline packages/better-auth/src/plugins/two-factor/index.ts]: the schema shouldn't be imported. It should be passed directly here to keep it consistent with other endpoints.

File: packages/better-auth/src/plugins/two-factor/index.ts [hunk 5]
```diff
@@ -73,22 +105,83 @@ export const twoFactor = (options?: TwoFactorOptions) => {
...
+								message: BASE_ERROR_CODES.INVALID_PASSWORD,
+							});
+						}
+					} else {
+						// Email OTP verification
+						const verificationValue =
+							await ctx.context.internalAdapter.findVerificationValue(
+								`2fa-otp-${user.id}`,
+							);
+							console.log('Verification attempt:', {
```
Comment (Bekacru) [inline packages/better-auth/src/plugins/two-factor/index.ts]: let's remove log

**Review by Bekacru — CHANGES_REQUESTED** [review]

**Comment by billdotdev** [comment]
Hi, I think you can't `sendOtp` without enabling two factor because `send2FaOTP` checks for `user.twoFactorEnabled` and requires a twoFactor row for the user?
```
// Using email OTP verification (for social login users)
// First send OTP
await client.twoFactor.sendOtp()

// Then enable 2FA with the received OTP
const { data } = await authClient.twoFactor.enable({
    otp: "received-otp"
})
```

**Comment by michaelhays** [comment]
Anything we can do to help push this forward? This would be huge for our team :pray:

**Comment by michaelhays** [comment]
Hey @Bekacru, any chance this will be reviewed and merged soon? Is there anything else missing?

All good if not, just wondering because we've had a hacky workaround that we've put off deploying in the hopes that this will be ready soon. But if it seems like this might take a few weeks longer before it's merged, then we'll go ahead and use the workaround.

**Comment by Bekacru** [comment]
@michaelhays we currently don't support 2fa for social logins. They automatically skip 2fa step. We're thinking of introducing that now this is possible. That's why we're waiting on this PR. It'll be released in 1.3 hopefully next friday.

**Comment by michaelhays** [comment]
Got it! Thanks for the context 🙏

**Comment by iatomic1** [comment]
Hey @Bekacru any updates on this?

**Comment by csyedbilal** [comment]
@Bekacru any update?

**Comment by luksch42** [comment]
@Bekacru Now that 1.3 is in beta, is anything still holding things up?

**Comment by ping-maxwell** [comment]
@helloworld9912 Can we fix the merge conflicts? thanks
